feat(python): Support arithmetic between Series with dtype list

itamarst opened the pull request:
Fixes #9188 
Fixes #14541

itamarst commented:
Ooh that's a real test failure on Windows, and it's a _fun_ test failure.

itamarst commented:
Ah: https://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine

ritchie46 commented on crates/polars-core/src/series/arithmetic/borrowed.rs:
```diff
@@ -171,6 +171,50 @@ impl NumOpsDispatchInner for FixedSizeListType {
+        polars_ensure!(self.len() == rhs.len(), InvalidOperation: "can only do arithmetic operations on Series of the same size; got {} and {}", self.len(), rhs.len());
+
+        let mut result = self.clear();
+        let combined = self.amortized_iter().zip(rhs.list()?.amortized_iter()).map(|(a, b)| {
```
I really think we should work on leaves and later add the `offset`s. I want to multiply the base arrays without the nesting and later restore the proper nesting structure.

itamarst replied:
I did that by adding the leaf values array directly, unchanged, and ... then I realized that as soon as you have nulls adding those two together is no longer necessarily correct:

```python
>>> df = pl.Series([[1, 2], None, [3, 4]])
>>> df2 = pl.Series([None, [1, 2], [3, 4]])
>>> df + df2
shape: (3,)
Series: '' [list[i64]]
[
        [2, 4]
        null
        [6, 8]
]
```

In the underlying Arrow representation there's an (Arrow) values array of length 4 for each... but the offsets are different when the nulls don't line up.

That result is wrong. So if there are any nulls you fall back to something using `amortized_iter()`. I'm fairly certain the code as originally submitted also handles nulls incorrectly, but in a different way... so first I'm going to write a bunch of tests.

Currently thinking there will be two code paths, a fast path for no nulls (adds the underlying values arrays), and a slow path for when there are nulls (the logic in the original PR, more or less).

itamarst commented:
(I'm sick, will eventually get back to this when recovered.)

ritchie46 commented:
Take care.

itamarst commented:
I merged forward and fixed compilation errors, so _now_ it should be ready for review.

ritchie46 commented on Cargo.toml:
```diff
@@ -76,6 +76,7 @@ recursive = "0.1"
 regex = "1.9"
 reqwest = { version = "0.12", default-features = false }
 ryu = "1.0.13"
+scopeguard = "1.2.0"
```
We want to reduce dependencies already. I don't think its worth a whole new dependency.

itamarst replied:
OK

itamarst replied:
This is already a transitive dependency, for what it's worth, since it's a very popular library.

ritchie46 commented on crates/polars-core/src/series/arithmetic/borrowed.rs:
```diff
@@ -171,6 +175,168 @@ impl NumOpsDispatchInner for FixedSizeListType {
+            rhs.len()
+        );
+
+        thread_local! {
```
Why do we deal with `thread_local` and statics here?

itamarst replied:
If this makes sense I'll add an explanatory comment.

itamarst replied:
And I should add, recursive in a way that doesn't allow passing arguments, since there's a whole bunch of layers in between.

ritchie46 replied:
We should pass the the arguments. Going thread_local isn't an option I want for arithmetic, and to be fair almost never in the codebase.. I also think it is incorrect. A thread can be halted in the middle of this work and then another rayon task can start on this thread and overwrite the thread local state.

itamarst replied:
This would involve modifying a bunch of more general traits, e.g. `PrivateSeries::add_to`, to have an extra argument. Which would be somewhat intrusive, but possible I guess.

Another option is just to stick to the slow path only.

itamarst replied:
A third option is that I'm wrong and recursion isn't necessary, or could be done in a way that bypasses the more general-purpose APIs. So I'll try to see if I can figure that out, and report back.

itamarst replied:
OK, figured it out.

ritchie46 commented on crates/polars-core/src/series/arithmetic/borrowed.rs:
```diff
@@ -171,6 +175,168 @@ impl NumOpsDispatchInner for FixedSizeListType {
+                    a.len(),
+                    b.len()
+                );
+                op(a, b).and_then(|s| s.implode()).map(Series::from)
```
I think we can keep `ListChunked` here. Instead of going int `Series` and then unpacking `List`  later.

itamarst replied:
OK will try to fix that.

itamarst commented:
I have addressed the 3 review comments above.

ritchie46 commented on crates/polars-core/src/series/arithmetic/borrowed.rs:
```diff
@@ -171,6 +173,167 @@ impl NumOpsDispatchInner for FixedSizeListType {
     }
 }
 
+/// Given an ArrayRef with some primitive values, wrap it in list(s) until it
```
Can we create a submodule `list_borrowed`? I think it is better to have this code local and this module a bit smaller.

ritchie46 reviewed (commented):
Alright @itamarst. I think this is the right approach. I have left some comments.

ritchie46 commented:
P.S. Can you make sure you rebase? There has been a huge refactor in between.

eitsupi commented:
Perhaps this is a feature addition to Rust so the PR title needs to be changed?

itamarst commented:
I have (hopefully) addressed the review comments.

itamarst commented:
Note I just added better test coverage, which caught a bug in expression type casting, which I fixed but maybe in the wrong place. See [commit 85cc6dd above.](https://github.com/pola-rs/polars/pull/17823/commits/85cc6dd054a299945e0951a040ddb2a1184565c6)

itamarst commented on crates/polars-expr/src/expressions/cast.rs:
```diff
@@ -13,7 +13,14 @@ pub struct CastExpr {
 impl CastExpr {
     fn finish(&self, input: &Series) -> PolarsResult<Series> {
-        input.cast_with_options(&self.dtype, self.options)
+        let dtype = if matches!(input.dtype(), DataType::List(_)) && !self.dtype.is_nested() {
```
Question for reviewer: this seems like the wrong place to do it. Is there a better place?

ritchie46 commented:
Thanks @itamarst. This is a great addition. :+1:
